### PR TITLE
fix: detect workspace package manager

### PR DESCRIPTION
**Pull request by AmirSa12**

### CODE REVIEW

**Comment by AmirSa12** [comment]
resolves #129

File: packages/cli/index.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ import pkg from "./package.json" with { type: "json" };
 import { isBinaryFile } from "isbinaryfile";
 import { readPackageJSON, writePackageJSON } from "pkg-types";
 import { createDefaultTemplate } from "./template";
+import { detectPackageManager } from "../backend/server/utils/markdown";
```
Comment (Aslemammad) [inline packages/cli/index.ts:23]: why is it in markdown utils? I don't think it should be there, let's have it inlined in this file!

File: packages/backend/server/routes/publish.post.ts [hunk 2]
```diff
@@ -7,19 +7,22 @@ import { randomUUID } from "uncrypto";
...
     "sb-run-id": runIdHeader,
     "sb-key": key,
     "sb-shasums": shasumsHeader,
     "sb-comment": commentHeader,
     "sb-compact": compactHeader,
+    "sb-package-manager": packageManagerHeader,
   } = getHeaders(event);
   const compact = compactHeader === "true";
   const comment: Comment = (commentHeader ?? "update") as Comment;
+  const packageManager = packageManagerHeader || detectPackageManager(".");
```
Comment (Aslemammad) [inline packages/backend/server/routes/publish.post.ts:25]: This is the server, how would we detect the user's packageManager from the server?

File: packages/backend/server/routes/publish.post.ts [hunk 3]
```diff
@@ -259,6 +265,19 @@ export default eventHandler(async (event) => {
   };
 });
 
+function detectPackageManager(rootPath: string): "npm" | "pnpm" {
```
Comment (Aslemammad) [inline packages/backend/server/routes/publish.post.ts:268]: This should be in the cli and not in the server!

File: packages/backend/server/utils/markdown.ts [hunk 4]
```diff
@@ -63,7 +65,7 @@ export function generatePullRequestPublishMessage(
         `<b>${packageName}</b>`,
         `
 \`\`\`
-npm i ${refUrl}
+${packageManager} ${packageManager === "npm" ? "i" : "add"} ${refUrl}
```
Comment (Aslemammad) [inline packages/backend/server/utils/markdown.ts:68]: let's just put `i` for both npm and pnpm

File: packages/cli/index.ts [hunk 5]
```diff
@@ -277,6 +277,20 @@ const main = defineCommand({
             }
           }
 
+          function detectPackageManager(rootPath: string): "npm" | "pnpm" {
```
Comment (Aslemammad) [inline packages/cli/index.ts:280]: put this method at the end of the file! like other methods.

**Review by Aslemammad — CHANGES_REQUESTED** [review]
Good progress!

File: packages/cli/index.ts [hunk 6]
```diff
@@ -403,3 +407,16 @@ async function hasPackageJson(p: string) {
     return false;
   }
 }
+
+function detectPackageManager(rootPath: string): "npm" | "pnpm" {
```
Comment (Aslemammad) [inline packages/cli/index.ts:411]: let's use what @sadeghbarati suggested or detect package manager by egoist!

File: packages/backend/server/utils/markdown.ts [hunk 7]
```diff
@@ -21,7 +22,7 @@ export function generateCommitPublishMessage(
         `<b>${packageName}</b>`,
         `
 \`\`\`
-npm i ${shaUrl}
+${packageManager} ${packageManager === "npm" ? "i" : "add"} ${shaUrl}
```
Comment (Aslemammad) [inline packages/backend/server/utils/markdown.ts:25]: let's have an object like this

```
{
npm: "i",
pnpm: "add",
yarn: "add"
```
and then based on what the package manager is, we choose!

File: packages/backend/server/routes/publish.post.ts [hunk 8]
```diff
@@ -10,16 +10,17 @@ import { generateTemplateHtml } from "~/utils/template";
...
     "sb-run-id": runIdHeader,
     "sb-key": key,
     "sb-shasums": shasumsHeader,
     "sb-comment": commentHeader,
     "sb-compact": compactHeader,
+    "sb-package-manager": packageManagerHeader,
   } = getHeaders(event);
   const compact = compactHeader === "true";
   const comment: Comment = (commentHeader ?? "update") as Comment;
+  const packageManager = packageManagerHeader || "npm";
```
Comment (Aslemammad) [inline packages/backend/server/routes/publish.post.ts:23]: let's have a package manager of type PackageManager with possible values (npm, pnpm and yarn)

**Review by Aslemammad — CHANGES_REQUESTED** [review]
Amazing progress! love it.

File: package.json [hunk 9]
```diff
@@ -39,6 +39,7 @@
     }
   },
   "dependencies": {
+    "detect-package-manager": "^3.0.2",
```
Comment (Aslemammad) [inline package.json]: We have to put it in cli package json, not here!

File: packages/backend/server/utils/markdown.ts [hunk 10]
```diff
@@ -1,12 +1,19 @@
 import { abbreviateCommitHash } from "@pkg-pr-new/utils";
 import { WorkflowData } from "../types";
 
+const packageCommands: Record<string, string> = {
```
Comment (Aslemammad) [inline packages/backend/server/utils/markdown.ts:4]: `Record<PackageManager, string>`

File: packages/backend/server/utils/markdown.ts [hunk 11]
```diff
@@ -1,12 +1,19 @@
...
+  yarn: "add",
+};
+
 export function generateCommitPublishMessage(
   origin: string,
   templates: Record<string, string>,
   packages: string[],
   workflowData: WorkflowData,
   compact: boolean,
+  packageManager: string,
```
Comment (Aslemammad) [inline packages/backend/server/utils/markdown.ts:16]: Let's use `PackageManager` type here!

File: packages/backend/server/utils/markdown.ts [hunk 12]
```diff
@@ -47,7 +54,8 @@ export function generatePullRequestPublishMessage(
   compact: boolean,
   checkRunUrl: string,
   codeflow: boolean,
-  base: "sha" | "ref"
+  packageManager: string,
```
Comment (Aslemammad) [inline packages/backend/server/utils/markdown.ts:57]: same here!

File: packages/cli/index.ts [hunk 13]
```diff
@@ -277,6 +278,11 @@ const main = defineCommand({
             }
           }
 
+          const detectPackageManager = detect();
```
Comment (Aslemammad) [inline packages/cli/index.ts:281]: just await, no need for `.then`

File: packages/cli/index.ts [hunk 14]
```diff
@@ -403,3 +410,16 @@ async function hasPackageJson(p: string) {
     return false;
   }
 }
+
+// function detectPackageManager(rootPath: string): "npm" | "pnpm" {
```
Comment (Aslemammad) [inline packages/cli/index.ts:414]: Let's remove these comments!

File: packages/backend/server/routes/publish.post.ts [hunk 15]
```diff
@@ -7,19 +7,22 @@ import { randomUUID } from "uncrypto";
 import { setItemStream, useTemplatesBucket } from "~/utils/bucket";
 import { useOctokitInstallation } from "~/utils/octokit";
 import { generateTemplateHtml } from "~/utils/template";
+import type { PackageManager } from "@pkg-pr-new/utils";
```
Comment (Aslemammad) [inline packages/backend/server/routes/publish.post.ts:10]: If this type is only used in this file, then let's inline it here and remove it from utils!

**Review by Aslemammad — CHANGES_REQUESTED** [review]

File: packages/cli/index.ts [hunk 16]
```diff
@@ -277,6 +278,8 @@ const main = defineCommand({
             }
           }
 
+          const detectPackageManager = detect();
```
Comment (Aslemammad) [inline packages/cli/index.ts:281]: just const packageManager = await detect() is enough

**Review by Aslemammad — CHANGES_REQUESTED** [review]

**Review by Aslemammad — APPROVED** [review]

**Comment by Aslemammad** [comment]
Thank you so much for the hard work!

**Pull request merged**